### PR TITLE
feat: add an extra method to (Checked)Euclid trait to get both quotien and rem

### LINKED ISSUE #290: add a quotien_and_rem_euclid method to Euclid trait
Sometimes you need both the quotient and the remainder.

With the current state of the trait you would have to call both `div_euclid` and `rem_euclid`.
But part of the computation done by each one of those methods can be shared.
See:
```rust
impl Euclid for f32 {
    #[inline]
    fn div_euclid(&self, v: &f32) -> f32 {
        let q = <f32 as crate::float::FloatCore>::trunc(self / v);
        if self % v < 0.0 {
            return if *v > 0.0 { q - 1.0 } else { q + 1.0 };
        }
        q
    }

    #[inline]
    fn rem_euclid(&self, v: &f32) -> f32 {
        let r = self % v;
        if r < 0.0 {
            r + <f32 as crate::float::FloatCore>::abs(*v)
        } else {
            r
        }
    }
}
```

can become
```rust
#[inline]
    fn div_and_rem_euclid(&self, v: &f32) -> (f32, f32) {
        let q = <f32 as crate::float::FloatCore>::trunc(self / v); 
        let r = self % v;
        if r < 0.0 {
            (if *v > 0.0 { q - 1.0 } else { q + 1.0 }, r + <f32 as crate::float::FloatCore>::abs(*v))
        } else {
            (q, r)
        }
    }
````
And you spare one computation of `self % v`.
This is a really small optimization for f32, but with the kind of big field elements I'm working with, it actually a pretty significant hit.
The method can be defined with a default impl:
```rust
fn div_and_rem_euclid(&self, v: &f32) -> (f32, f32) {
        (self.div_euclid(), self.rem_euclid())
}
```

Which has not optimization nor drawback for people using it, but can be overwritten in order to achieve better performances.

Wdyt?

### PR DESCRIPTION
**Pull request by tdelabro**
closes #290

### CODE REVIEW

File: src/ops/euclid.rs [hunk 1]
```diff
@@ -46,6 +46,14 @@ pub trait Euclid: Sized + Div<Self, Output = Self> + Rem<Self, Output = Self> {
     /// assert_eq!(Euclid::rem_euclid(&-a, &-b), 1);
     /// ```
     fn rem_euclid(&self, v: &Self) -> Self;
+
+    /// Returns both the euclidian division quotien and remainer
+    ///
+    /// By default, it internaly calls both `Euclid::div_euclid` and `Euclid::rem_euclid`,
+    /// but it can be overwritten in order to implement some optimization.
```
Comment (cuviper) [inline src/ops/euclid.rs:53]:
```suggestion
    /// By default, it internally calls both `Euclid::div_euclid` and `Euclid::rem_euclid`,
    /// but it can be overridden in order to implement some optimization.
```

File: src/ops/euclid.rs [hunk 2]
```diff
@@ -174,6 +182,14 @@ pub trait CheckedEuclid: Euclid {
     /// Finds the euclid remainder of dividing two numbers, checking for underflow, overflow and
     /// division by zero. If any of that happens, `None` is returned.
     fn checked_rem_euclid(&self, v: &Self) -> Option<Self>;
+
+    /// Returns both the checked euclidian division quotien and remainer
```
Comment (cuviper) [inline src/ops/euclid.rs:186]:
```suggestion
    /// Returns both the quotient and remainder from checked Euclidean division.
```

File: src/ops/euclid.rs [hunk 3]
```diff
@@ -174,6 +182,14 @@ pub trait CheckedEuclid: Euclid {
...
+    ///
+    /// By default, it internaly calls both `Euclid::div_euclid` and `Euclid::rem_euclid`,
+    /// but it can be overwritten in order to implement some optimization.
```
Comment (cuviper) [inline src/ops/euclid.rs:189]:
```suggestion
    /// By default, it internally calls both `Euclid::checked_div_euclid` and `Euclid::checked_rem_euclid`,
    /// but it can be overridden in order to implement some optimization.
```

**Review by cuviper — CHANGES_REQUESTED** [review]
I suggested some fixes for the documentation. I would also appreciate some tests, and doc-examples are an easy way to both demonstrate the feature and test it.

I think we should drop "with" from the name, just `div_rem_euclid` and `checked_div_rem_euclid`. This will be more similar to [`Integer::div_rem`](https://docs.rs/num-integer/latest/num_integer/trait.Integer.html#tymethod.div_rem).

File: src/ops/euclid.rs [hunk 4]
```diff
@@ -46,6 +46,14 @@ pub trait Euclid: Sized + Div<Self, Output = Self> + Rem<Self, Output = Self> {
     /// assert_eq!(Euclid::rem_euclid(&-a, &-b), 1);
     /// ```
     fn rem_euclid(&self, v: &Self) -> Self;
+
+    /// Returns both the euclidian division quotien and remainer
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/ops/euclid.rs:50]: done

File: src/ops/euclid.rs (hunk 1 above)
Comment (tdelabro) [reply src/ops/euclid.rs:53]: done

File: src/ops/euclid.rs (hunk 2 above)
Comment (tdelabro) [reply src/ops/euclid.rs:186]: donde

File: src/ops/euclid.rs (hunk 3 above)
Comment (tdelabro) [reply src/ops/euclid.rs:189]: done

File: src/ops/euclid.rs [hunk 5]
```diff
@@ -327,8 +373,6 @@ mod tests {
                     {
                         assert_eq!(CheckedEuclid::checked_div_euclid(&$t::min_value(), &-1), None);
                         assert_eq!(CheckedEuclid::checked_rem_euclid(&$t::min_value(), &-1), None);
-                        assert_eq!(CheckedEuclid::checked_div_euclid(&1, &0), None);
-                        assert_eq!(CheckedEuclid::checked_rem_euclid(&1, &0), None);
```
Comment (cuviper) [inline src/ops/euclid.rs:331]: Why was this removed?

**Comment by tdelabro** [comment]
@cuviper Ready for review

File: src/ops/euclid.rs [hunk 6]
```diff
@@ -46,6 +46,27 @@ pub trait Euclid: Sized + Div<Self, Output = Self> + Rem<Self, Output = Self> {
     /// assert_eq!(Euclid::rem_euclid(&-a, &-b), 1);
     /// ```
     fn rem_euclid(&self, v: &Self) -> Self;
+
+    /// Returns both the quotient and remainder from Euclidean division.
+    ///
+    /// By default, it internally calls both `Euclid::div_euclid` and `Euclid::rem_euclid`,
+    /// but it can be overidden in order to implement some optimization.
```
Comment (cuviper) [inline src/ops/euclid.rs:53]:
```suggestion
    /// but it can be overridden in order to implement some optimization.
```

**Comment by cuviper** [comment]
Thanks!

bors r+
